ta/pkcs11/src/pkcs11_token.c

```diff
@@ -230,3 +230,59 @@ uint32_t entry_ck_token_info(uint32_t ptypes, TEE_Param *params)
 
 	return PKCS11_CKR_OK;
 }
+
+uint32_t entry_ck_token_mecha_ids(uint32_t ptypes, TEE_Param *params)
+{
+	const uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INOUT,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_MEMREF_OUTPUT,
+						TEE_PARAM_TYPE_NONE);
+	TEE_Param *ctrl = &params[0];
+	TEE_Param *out = &params[2];
+	uint32_t rv = 0;
+	struct serialargs ctrlargs = { };
+	uint32_t token_id = 0;
+	struct ck_token *token = NULL;
+	size_t count = 0;
+	size_t __maybe_unused n = 0;
+
+	if (ptypes != exp_pt)
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	serialargs_init(&ctrlargs, ctrl->memref.buffer, ctrl->memref.size);
+
+	rv = serialargs_get(&ctrlargs, &token_id, sizeof(token_id));
+	if (rv)
+		return rv;
+
+	if (serialargs_remaining_bytes(&ctrlargs))
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	token = get_token(token_id);
+	if (!token)
+		return PKCS11_CKR_SLOT_ID_INVALID;
+
+	count = get_supported_mechanisms(NULL, 0);
+
+	if (out->memref.size < count * sizeof(uint32_t)) {
+		out->memref.size = count * sizeof(uint32_t);
+		return PKCS11_CKR_BUFFER_TOO_SMALL;
+	}
+
+	(void)get_supported_mechanisms(out->memref.buffer, count);
+	out->memref.size = sizeof(uint32_t) * count;
+
+#ifdef DEBUG
```

Comment:
``` C
#if TRACE_LEVEL >= TRACE_DEBUG
```
This debug thing look a bit strange. Do you need it? If so, I'd prefer a helper function next to `get_supported_mechanisms()` which would parse the buffer.

Reply:
ok